RFC: Create a module for picolibc

keith-packard opened the pull request:
Rather than depending on the toolchain, create a module for picolibc and build it along with Zephyr. This has a few hacks to make it work; I'm mostly curious if this seems like a reasonable direction for picolibc integration. I've set it up to just use 'meson' to configure and build picolibc for now.

Closes #44143

keith-packard committed:
lib/libc: Add picolibc support (aarch32, aarch64 and RISC-V) [v21]

Picolibc is a fork of newlib designed and tested on embedded systems. It
offers a smaller memory footprint (both ROM and RAM), and native TLS
support, which uses the Zephyr TLS support.

By default, the full printf version is included in the executable, which
includes exact floating point and long long input and output. A
configuration option has been added to switch to the integer-only
version (which also omits long long support).

Here are some size comparisons using qemu-cortex-m3 and this application
(parameters passed to printf to avoid GCC optimizing it into puts):

void main(void)
{
    printf("Hello World! %s %d\n", CONFIG_BOARD, 12);
}

                       FLASH    SRAM
    minimal             8696    3952
    picolibc int        7600    3960
    picolibc float     12304    3960
    newlib-nano int    11696    4128
    newlib-nano float  30516    4496
    newlib             34800    6112

---

v2:
	Include picolibc-tls.ld

v3:
	Document usage in guides/c_library.rst and
	getting_started/toolchain_other_x_compilers.rst

v4:
	Lost the lib/libc/picolibc directory somehow!

v5:
	Add PICOLIBC_ALIGNED_HEAP_SIZE configuration option.
	Delete PICOLIBC_SEMIHOST option support code

v6:
	Don't allocate static RAM for TLS values; TLS
	values only need to be allocated for each thread.

v7:
	Use arm coprocessor for TLS pointer storage where supported for
	compatibility with the -mtp=cp15 compiler option (or when the
	target cpu type selects this option)

	Add a bunch of tests

	Round TLS segment up to stack alignment so that overall stack
	remains correctly aligned

	Add aarch64 support

	Rebase to upstream head

v8:
	Share NEWLIB, NEWLIB_NANO and PICOLIBC library configuration
	variables in a single LIBC_PARTITIONS variable instead of
	having separate PICOLIBC_PART and NEWLIB_PART variables.

v9:
	Update docs to reference pending sdk-ng support for picolibc

v10:
	Support memory protection by creating a partition for
	picolibc shared data and any pre-defined picolibc heap.

v11:
	Fix formatting in arch/arm/core/aarch64/switch.S

v12:
	Remove TLS support from this patch now that TLS is upstream
	Require THREAD_LOCAL_STORAGE when using PICOLIBC for architectures
	that support it.

v13:
	Merge errno changes as they're only needed for picolibc.
	Adapt cmake changes suggested by Torsten Tejlmand Rasmussen

v14:
	Update to picolibc 1.7 and newer (new stdin/stdout/stderr ABI)

v15:
	Respond to comments from dcpleung:
	* switch kernel/errno to use CONFIG_LIBC_ERRNO instead of
          CONFIG_PICOLIBC
	* Add comment to test/lib/sprintf as to why the %n test
	  was disabled for picolibc.

v16:
	Switch picolibc to a module built with Zephyr. This eliminates
	toolchain dependencies and allows compiler settings for Zephyr
	to also be applied to picolibc.

v17:
	Provide Zephyr-specific 'abort' implementation.
	Support systems with MMU

v18:
	Allow use of toolchain picolibc version.

v19:
	Use zephyr/ for zephyr headers

v20:
	Add locking
	Use explicit commit for picolibc module

v21:
	Create PICOLIBC_SUPPORTED config param. Set on arc, arm, arm64,
	mips and riscv architectures.

Signed-off-by: Keith Packard <keithp@keithp.com>

carlescufi commented:
Thanks @keith-packard, would you mind creating an issue as described [here](https://docs.zephyrproject.org/latest/contribute/external.html) so we can proceed with the creation of a mirror repo after a TSC vote?

carlescufi commented:
Please ignore the "Manifest" job failure, that's a bug on the bot side.

stephanosio commented:
Just to give some context to those who have not been involved in this conversation earlier:
https://github.com/picolibc/picolibc/issues/177#issuecomment-1073815655

stephanosio reviewed (commented):
I have listed the configuration requirements for the picolibc to be able to replace the current Zephyr minimal libc in the following comment:
https://github.com/picolibc/picolibc/issues/177#issuecomment-1076039486

In general, we want more granular control over the internals of the picolibc as described in that comment.

Also, we would want to be able to choose (via Kconfig) the components of the picolibc to compile so that, when we build the Zephyr applications, we do not build the entire picolibc and unreasonably increase the build time. Note that this is important because, in case we decide to go through with replacing the Zephyr minimal libc with the picolibc, the picolibc will become the default libc for Zephyr.

keith-packard commented:
I've mirrored the existing picolibc configuration options that are relevant to Zephyr into the Zephyr C Library configuration menu and apply those using the CMakeLists.txt provided on the zephyr branch of the picolibc repository. This ensures that the build mechanism used for picolibc is not exposed in the Zephyr repository, while still offering complete customization of the picolibc build.

keith-packard commented:
I pushed some preliminary cmake bits to the [cmake branch of picolibc](https://github.com/picolibc/picolibc/tree/cmake). I'm a bit stuck though; I need to set some per-source compiler options and tried using this function:
```
function(picolibc_sources_flags flags)
  set(orig_sources ${ARGN})
  picolibc_sources(${orig_sources})
  foreach(source IN LISTS orig_sources)
    message(STATUS "set flags for ${CMAKE_CURRENT_SOURCE_DIR}/${source} to ${flags}")
    set_source_files_properties(${source}
      DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
      PROPERTIES COMPILE_FLAGS "${flags}")
  endforeach()
endfunction()
```
but I can't get it to work. I tried open-coding the call to set_source_files_properties and that also failed. I'm otherwise pretty close to building a working C library.

keith-packard commented:
ok, I think I've got it building correctly, or at least can fix up any remaining issues. I'll see if I can't get zephyr to build tomorrow.

keith-packard commented:
And picolibc now builds with cmake as a module under zephyr, with the result working on arm32, riscv32 and riscv64. All of the changes for that were done in the picolibc repository on the zephyr branch, which has been pushed to https://github.com/zephyrproject-rtos/picolibc

keith-packard committed:
tests: Add picolibc tests

This adds picolibc-specific configurations for a couple of tests

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/sprintf: Skip UB-dependent tests on Picolibc

Passing NULL as the stream to stdio functions is "UB" (undefined behavior)
in the C standard. Picolibc crashes in this case, which is valid, but not
helpful. Skip these tests in this case.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/c_lib: Don't assume a particular PRNG from Picolibc [v2]

The random/rand tests assume the PRNG operates in a specific fashion,
which is not true when running Picolibc.

v2:
	Call ztest_test_skip for picolibc

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
I've spent some quality time updating this patch series. At this point, it runs nearly all of the twister tests against the arm architecture.
```
$ ./scripts/twister --ninja -a arm >& twister.log
```
The memory protection tests continue to fail, and I can't figure them out -- they fault when attempting to access the z_arm_tls_ptr value, but the section containing that variable is included in the thread's list of MPU regions.

The picolibc changes to integrate cmake and zephyr support are moving upstream there; the zephyr branch contains all of the patches necessary to build zephyr against that. Advice on how to structure the zephyr-specific changes would be appreciated, it seems like most of the zephyr-specific files could be moved in to the zephyr tree?

keith-packard commented:
And now I've cleaned up the Zephyr patch series so that it has all core Zephyr changes, followed by the Picolibc additions and finally changes to tests to make them work with Picolibc too. The last patch in this series changes the default C library from Minimal to Picolibc; I've included that to aid testing. With the patches here and the Zephyr branch from picolibc, which contains a few Picolibc fixes, cmake support and Zephyr configuration files, twister fails only on the memory protection tests which crash when the app tries to fetch the TLS pointer.

I've started working on RISC-V as well -- that also needs code to set the TP register when switching to user mode (there's a patch in the series above for 32-bit arm).

keith-packard committed:
tests/sprintf: Adjust expected output for 5e-324

The smallest denorm is shown as 5e-324 by picolibc.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
Ok, it looks like zephyr needs some TLS attention -- I've discovered numerous tests are failing when thread local storage is enabled, even when using the minimal C library. I've got patches for enough to make twister pass on -a arm and am checking -a riscv32 -a riscv64 to see how much more work is needed here.

This means that I believe picolibc support is now actually ready for final review -- the vast bulk of the patches in this series affect tests which need updating to support TLS. The actual picolibc changes are pretty minor. All of the necessary picolibc patches are moving upstream and are present on the zephyr branch.

stephanosio commented:
Rebased

stephanosio commented:
Will target 3.2.0 for this feature since the feature freeze for 3.1.0 is on this Friday.

stephanosio reviewed (commented):
@keith-packard I think it would be helpful to split this series into two parts:

* Essential commits required for introducing the picolibc support to Zephyr (this PR)
* Extraneous commits that are not absolutely needed for supporting the picolibc in Zephyr (another PR); for example:
    * [lib/os: Use picolibc printf as a replacement for printk](https://github.com/zephyrproject-rtos/zephyr/pull/44096/commits/90478fc42fe8203ee1f3b63f88c0108edddd54e9)
    * [lib/os: Replace cbvprintf and printfcb family when using picolibc [v3]](https://github.com/zephyrproject-rtos/zephyr/pull/44096/commits/8e3473bc3366f75f3c595b1cdaf15f0ee606f2fb)

Some of the "extraneous commits" may be controversial and will likely take some time for everyone to reach a consensus; so, it would be good to get them reviewed in a separate PR so as to not block the introduction of the picolibc itself.

keith-packard commented:
> @keith-packard I think it would be helpful to split this series into two parts:
> 
>     * Essential commits required for introducing the picolibc support to Zephyr (this PR)

Pushed to picolibc-module. I've also included the picolibc specific tests which don't depend on making picolibc the default C library, and the patches to those tests necessary to make them work.

>     * Extraneous commits that are not absolutely needed for supporting the picolibc in Zephyr (another PR); for example:
>       
>       * [lib/os: Use picolibc printf as a replacement for printk](https://github.com/zephyrproject-rtos/zephyr/pull/44096/commits/90478fc42fe8203ee1f3b63f88c0108edddd54e9)
>       * [lib/os: Replace cbvprintf and printfcb family when using picolibc [v3]](https://github.com/zephyrproject-rtos/zephyr/pull/44096/commits/8e3473bc3366f75f3c595b1cdaf15f0ee606f2fb)

I've pushed those to a 'picolibc-complete' branch; let's get picolibc merged and then I'll submit two separate PRs for those bits. 

> Some of the "extraneous commits" may be controversial and will likely take some time for everyone to reach a consensus; so, it would be good to get them reviewed in a separate PR so as to not block the introduction of the picolibc itself.

I hope not -- they both replicate existing APIs while making them smaller with better POSIX conformance. But, we'll see.

stephanosio commented on lib/libc/picolibc/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+zephyr_library()
+zephyr_library_sources(libc-hooks.c)
+
+if(CONFIG_PICOLIBC AND NOT CONFIG_PICOLIBC_USE_MODULE)
```
`if(CONFIG_PICOLIBC` is redundant because the `picolibc` subdirectory is only added when `CONFIG_PICOLIBC=y`.

keith-packard replied:
Fixed.

stephanosio replied:
It looks like this change was not pushed.

stephanosio commented on lib/libc/picolibc/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+
+  # Use picolibc provided with the toolchain
+
+  zephyr_compile_options(-D_POSIX_C_SOURCE=200809 --specs=picolibc.specs)
```
For `-D_POSIX_C_SOURCE=200809`:
```
zephyr_compile_definitions(_POSIX_C_SOURCE=200809)
```

For `--specs=picolibc.specs`:
```
  zephyr_link_libraries(-specs=picolibc.specs)
  zephyr_compile_options(-specs=picolibc.specs)
```

keith-packard replied:
Thanks, fixed.

stephanosio replied:
It looks like this change was not pushed.

stephanosio commented on lib/libc/picolibc/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+  zephyr_compile_options(-D_POSIX_C_SOURCE=200809 --specs=picolibc.specs)
+  zephyr_link_libraries(-T/dev/null --specs=picolibc.specs c -lgcc)
+  if(CONFIG_PICOLIBC_IO_FLOAT)
+    zephyr_compile_options(-DPICOLIBC_DOUBLE_PRINTF_SCANF)
+    zephyr_link_libraries(-DPICOLIBC_DOUBLE_PRINTF_SCANF)
```
```
zephyr_compile_definitions(PICOLIBC_DOUBLE_PRINTF_SCANF)
```

keith-packard replied:
Fixed.

stephanosio replied:
It looks like this change was not pushed.

stephanosio commented on lib/libc/picolibc/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+    zephyr_compile_options(-DPICOLIBC_DOUBLE_PRINTF_SCANF)
+    zephyr_link_libraries(-DPICOLIBC_DOUBLE_PRINTF_SCANF)
+  else()
+    zephyr_compile_options(-DPICOLIBC_INTEGER_PRINTF_SCANF)
+    zephyr_link_libraries(-DPICOLIBC_INTEGER_PRINTF_SCANF)
```
```
zephyr_compile_definitions(PICOLIBC_INTEGER_PRINTF_SCANF)
```

keith-packard replied:
Fixed

stephanosio replied:
It looks like this change was not pushed.

stephanosio commented on west.yml:
```diff
@@ -199,6 +199,9 @@ manifest:
       path: modules/lib/openthread
+    - name: picolibc
+      path: modules/lib/picolibc
+      revision: zephyr
```
This needs to be pointing to a specific commit.

keith-packard replied:
Thanks; I was leaving it pointing to a branch so I didn't need to update zephyr when I made changes in picolibc. Now that there's a released version of picolibc with the required changes, I'll switch to a commit instead. I assume you don't want a tag name here?

stephanosio replied:
> I assume you don't want a tag name here?

While specifying a tag name should work, commit SHA would be preferred.

[We are in the process of cleaning up the upstream module repositories of unneeded branches (and tags)](https://github.com/zephyrproject-rtos/zephyr/issues/45241), so using commit SHA would be required as part of that too.

keith-packard replied:
Fixed. Points to the 1.7.7 commit sha.

stephanosio commented on tests/lib/c_lib/src/main.c:
```diff
@@ -1065,11 +1068,13 @@ ZTEST(test_c_lib, test_time)
  */
 ZTEST(test_c_lib, test_rand)
 {
+#ifndef CONFIG_PICOLIBC
```
```
#else
	ztest_test_skip();
```

stephanosio commented on tests/lib/c_lib/src/main.c:
```diff
@@ -1079,6 +1084,7 @@ ZTEST(test_c_lib, test_rand)
  */
 ZTEST(test_c_lib, test_srand)
 {
+#ifndef CONFIG_PICOLIBC
```
```
#else
	ztest_test_skip();
```

stephanosio commented on tests/lib/c_lib/src/main.c:
```diff
@@ -1109,6 +1116,7 @@ ZTEST(test_c_lib, test_srand)
  */
 ZTEST(test_c_lib, test_rand_reproducibility)
 {
+#ifndef CONFIG_PICOLIBC
```
```
#else
	ztest_test_skip();
```

stephanosio commented:
@keith-packard How do we ensure thread-safety with the picolibc (e.g. calling `malloc` from multiple threads)?

At first glance, it seems like the picolibc functions will not be thread safe because the `libc-hooks.c` for it is missing the retargetable locking interface implementation, is that correct?

keith-packard commented on include/zephyr/sys/libc-hooks.h:
```diff
@@ -57,7 +57,21 @@ extern struct k_mem_partition z_malloc_partition;
+ *    aligned base addreess after the `_end` linker symbol, through to the end
+ *    of system RAM.
+ */
+#if (!defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT) || \
+(defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT) && \
+CONFIG_PICOLIBC_ALIGNED_HEAP_SIZE))
```
The code style checks disagree with you here; I suspect there's a bug in that.

keith-packard commented on include/zephyr/sys/libc-hooks.h:
```diff
@@ -67,7 +81,7 @@ extern struct k_mem_partition z_malloc_partition;
 #endif
 
 #if defined(CONFIG_NEWLIB_LIBC) || defined(CONFIG_STACK_CANARIES) || \
-    defined(CONFIG_NEED_LIBC_MEM_PARTITION)
+defined(CONFIG_PICOLIBC) || defined(CONFIG_NEED_LIBC_MEM_PARTITION)
```
same issue -- code style requires no indentation on these continuation lines.

keith-packard commented:
> @keith-packard How do we ensure thread-safety with the picolibc (e.g. calling `malloc` from multiple threads)?
> 
> At first glance, it seems like the picolibc functions will not be thread safe because the `libc-hooks.c` for it is missing the retargetable locking interface implementation, is that correct?

Yes. I borrowed the newlib hooks (amended for picolibc) in the most recent update.

keith-packard commented:
> Pushed to picolibc-module. I've also included the picolibc specific tests which don't depend on making picolibc the default C library, and the patches to those tests necessary to make them work.

I've also pushed a picolibc-fixes branch, which has all of the test/sample fixes needed to run the whole test suite using picolibc (on supported platforms).

Finally, I pushed picolibc-printk and picolibc-cbprintf which have the patches to enable those two features on top of the picolibc-module commit. They conflict as they both touch libc-hooks.c, but they are otherwise independently reviewable.

stephanosio commented on CMakeLists.txt:
```diff
@@ -934,6 +943,9 @@ if(CONFIG_USERSPACE)
   if(CONFIG_NEWLIB_LIBC_NANO)
     set(NEWLIB_PART -l libc_nano.a z_libc_partition -l libm_nano.a z_libc_partition)
   endif()
+  if(CONFIG_PICOLIBC)
+    set(NEWLIB_PART -l libc.a z_libc_partition)
+  endif()
```
Thanks for clarifying. Yes, it would be good to have that patch back in this series.

keith-packard replied:
Sounds good. I've rebased and amended this series; will push (and resolve) when local tests are complete.

keith-packard replied:
Local tests completed; found a couple more bugs (PRs posted), then pushed this branch again.

keith-packard committed:
tests/c_lib: Ignore stringop-overflow warning

One test reads too few bytes from the source to ensure nul-termination of
the result which generates a compiler warning.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
soc/it8xxx2: Place arithmetic funcs in .text section

Placing these in the __ram_code section generates a relocation error when
building with toolchain version 0.14.1, moving them back to .text fixes
that (presumably with a performance penalty).

Signed-off-by: Keith Packard <keithp@keithp.com>

andyross commented:
Seeing this go by and nodding every time.  Not to step too far into the project governance hole, but... what exactly are we waiting for here?  This is more libc engagement and development than we seen in... basically forever.  What's preventing us from just pulling the trigger and merging, with an eye to making picolibc default?

stephanosio commented:
> what exactly are we waiting for here?

Not much really. All requested changes seem to have been addressed, the TSC has already approved this, so nothing from the governance side of things either.

It should be good to go once the CI is green (there is a failure on `hifive_unleashed` for some reason):
https://github.com/zephyrproject-rtos/zephyr/runs/6956910072?check_suite_focus=true#step:12:1436

I will revisit this PR for review some time this week.

stephanosio commented on lib/libc/picolibc/libc-hooks.c:
Zephyr provides `CODE_UNREACHABLE` as part of the toolchain abstraction layer.

keith-packard replied:
and CODE_UNREACHABLE was already used a few lines above this. Fixed.

stephanosio commented on tests/lib/mem_alloc/testcase.yaml:
```diff
@@ -50,3 +50,9 @@ tests:
     - malloc
     - memalloc_max
     - reallocarray
+  libraries.libc.picolibc:
```
`libraries.libc.picolibc.mem_alloc`

keith-packard replied:
Fixed.

keith-packard commented:
> It should be good to go once the CI is green (there is a failure on `hifive_unleashed` for some reason): https://github.com/zephyrproject-rtos/zephyr/runs/6956910072?check_suite_focus=true#step:12:1436

Not sure how I can help with this -- I don't have an unleashed board any more. This used to pass, and I haven't changed the code in the patch for quite a while. Someone with the hardware will likely need to bisect the merge point against the last successful run, and that doesn't sound like much fun.

> I will revisit this PR for review some time this week.

Thanks. I'd like to get this off my todo list if possible.

nashif commented:
> > It should be good to go once the CI is green (there is a failure on `hifive_unleashed` for some reason): [zephyrproject-rtos/zephyr/runs/6956910072?check_suite_focus=true#step:12:1436](https://github.com/zephyrproject-rtos/zephyr/runs/6956910072?check_suite_focus=true#step:12:1436)
> 
> Not sure how I can help with this -- I don't have an unleashed board any more. This used to pass, and I haven't changed the code in the patch for quite a while. Someone with the hardware will likely need to bisect the merge point against the last successful run, and that doesn't sound like much fun.

this is happening in the renode simulator, so I do not think you need the board if you can deal with renode. Have seen issues with timing and renode in the past, could be just a false positive and something unique to renode. @mgielda @pgielda  @PiotrZierhoffer  can you please take a look?

keith-packard committed:
tests/mem_alloc: Make sure the "impossible" allocations actually are

The hifive_unleashed board is configured with 0xf0000000 bytes of memory,
so an allocation of 0x7fffffff should succeed. Instead, use a much larger
value to ensure that the allocation fails.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
> this is happening in the renode simulator, so I do not think you need the board if you can deal with renode. Have seen issues with timing and renode in the past, could be just a false positive and something unique to renode. @mgielda @pgielda @PiotrZierhoffer can you please take a look?

I added some debugging to the code so I could just let the CI run and see what happened. It turns out the hifive_unleashed is configured with a lot of memory (0xf000_0000 bytes). The test is requesting an allocation of 0x7fff_ffff bytes, which fits within this heap. As the test is calling calloc, it then calls memset on the result, which takes "a while". I assume the test is just timing out at that point. I've added a fix to the mem_alloc test to make sure allocations which "should" fail actually will.

keith-packard committed:
tests/log_stack: Increase stack usage for x86_64

The amount of stack used has changed for some reason (toolchain updates?),
use the amounts reported by the test when it failed.

Signed-off-by: Keith Packard <keithp@keithp.com>

stephanosio commented on tests/lib/mem_alloc/src/main.c:
```diff
@@ -31,6 +31,12 @@
 #include <errno.h>
 #include <time.h>
 
+#ifdef __PTRDIFF_MAX__
+#define TOO_BIG __PTRDIFF_MAX__
```
`__PTRDIFF_MAX__` is not standard and toolchain-specific.

`PTRDIFF_MAX`, which is part of the C standard from C99 and on, should be used.

keith-packard replied:
Thanks, I've updated the patch (without rebasing) to use PTRDIFF_MAX unconditionally after including stdint.h

stephanosio commented on tests/subsys/logging/log_stack/src/main.c:
```diff
@@ -133,9 +133,9 @@ ZTEST_SUITE(test_log_stack, NULL, NULL, NULL, NULL, after);
-#define SIMPLE_USAGE 1520
-#define HEXDUMP_USAGE 1504
-#define MORE_ARGS_USAGE 1584
+#define SIMPLE_USAGE 1552
+#define HEXDUMP_USAGE 1536
+#define MORE_ARGS_USAGE 1616
```
> The amount of stack used has changed for some reason (toolchain updates?),
use the amounts reported by the test when it failed.

This is likely from the main branch (same failure observed there too) -- we have been discussing whether this test should be disabled because the stack usage tends to change from time to time and it does not make sense to make the CI fail because of this (FYI @nordic-krch).

For now, this patch should do.

stephanosio reviewed (commented):
Thanks for investigating the issue. The fix looks reasonable to me.

Just one minor change in the fix.

PiotrZierhoffer commented:
@nashif @keith-packard I didn't follow the whole conversation, but it seems that the current code passes as expected because the previous version was supposed to fail. Is that correct?

Indeed, allocating this amount of memory in Renode can take a moment, but admittedly it's not a usual use case.

stephanosio commented on arch/Kconfig:
```diff
@@ -540,6 +545,12 @@ config ARCH_HAS_COHERENCE
 config ARCH_HAS_THREAD_LOCAL_STORAGE
 	bool
 
+config ARCH_HAS_PICOLIBC
+	bool
+	help
+	  Selected when the target architecture has validated support
+	  for picolibc.
```
IMHO, this would better handled via `depends on ARC || ARM || ARM64 || ...` from the `config PICOLIBC` because `ARCH_HAS_*` are for the availability of the architecture/ABI-specific kernel features, but no strong opinion in this particular instance since the picolibc will soon become an integral part of Zephyr.

keith-packard replied:
Hrm. What I want is a way to know when the picolibc-specific tests can be enabled; this replaced a pile of architecture tests in the testcase.yaml files. I guess I could stick a config PICOLIBC_SUPPORTS_ARCH in lib/libc/Kconfig? Name suggestions welcome, just want to make sure we have a single place which determines whether picolibc is supported or not, instead of littering them around the codebase.

stephanosio replied:
> What I want is a way to know when the picolibc-specific tests can be enabled

In that case, adding `PICOLIBC_SUPPORTED` or similar that has `default y` and `depends on ARC || ARM || ...` in `lib/libc/Kconfig` should do.

Also `config PICOLIBC` should depend on `PICOLIBC_SUPPORTED` since it would not be reasonable to allow users to select the picolibc for the archs that are not tested.

One question I have is: do we really need this? Is picolibc known to not work with any of the architectures currently supported in Zephyr? (i.e. do any of the tests fail on them?)

keith-packard replied:
Yeah, that looks better -- stick 'config ARCH_HAS_PICOLIBC' in lib/libc/Kconfig. Let me know if there's a better name for this.

stephanosio replied:
Since `config ARCH_HAS_PICOLIBC` is no longer in `arch/Kconfig`, `PICOLIBC_SUPPORTED` or `PICOLIBC_SUPPORTS_ARCH` would be better.

Though, this symbol may very well be generalised to indicate whether the picolibc with the current configuration is supported in general (i.e. not just for the different archs, but also for toolchains); in that sense, `PICOLIBC_SUPPORTED` could be a better choice than `PICOLIBC_SUPPORTS_ARCH`; but for now, either would be ok really.

stephanosio replied:
Also still wondering:

> One question I have is: do we really need this? Is picolibc known to not work with any of the architectures currently supported in Zephyr? (i.e. do any of the tests fail on them?)

keith-packard replied:
> Since `config ARCH_HAS_PICOLIBC` is no longer in `arch/Kconfig`, `PICOLIBC_SUPPORTED` or `PICOLIBC_SUPPORTS_ARCH` would be better.
> 
> Though, this symbol may very well be generalised to indicate whether the picolibc with the current configuration is supported in general (i.e. not just for the different archs, but also for toolchains); in that sense, `PICOLIBC_SUPPORTED` could be a better choice than `PICOLIBC_SUPPORTS_ARCH`; but for now, either would be ok really.

Sounds like `PICOLIBC_SUPPORTED` would be a better choice. I'll switch.

keith-packard replied:
> One question I have is: do we really need this? Is picolibc known to not work with any of the architectures currently supported in Zephyr? (i.e. do any of the tests fail on them?)

I tried x86_64 and there were some unexpected failures which I haven't debugged. Note that I'm doing much more significant testing off of the 'picolibc-fixes' branch, which switches the default C library to picolibc and then runs all of the tests, instead of only the subset included in the picolibc-module test. Small steps.

stephanosio replied:
> Note that I'm doing much more significant testing off of the 'picolibc-fixes' branch, which switches the default C library to picolibc and then runs all of the tests, instead of only the subset included in the picolibc-module test. Small steps.

Ok, thanks for clarifying. We can gradually address those issues after this PR is merged.

stephanosio reviewed (dismissed):
LGTM

Thanks for being patient and coping with the lengthy review process.

keith-packard committed:
tests/sprintf: Skip '%n' test on picolibc

Picolibc excludes %n support on purpose as it can introduce security issues
when the format string is vulnerable to manipulation.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
ok, I'll try to hold this version stable long enough to get a couple of approvals so we can just get 'er merged. If you have any comments about changes that are required, if they're relatively minor, let's plan on doing minor PRs after the merge to fix things up instead of delaying the merge.

andyross reviewed (approved):
Cool by me, certainly.  Nothing really impactful is happening in the Zephyr tree that I can see, mostly cleanup for existing weirdness.

stephanosio reviewed (approved):
LGTM

enjiamai commented:
Hi @keith-packard,  the commit of the picolibc repo seems like a incorrect reversion in the west.yml.  It will cause west update failed. Could you please help on correcting it? thanks.

```
199     - name: picolibc
200       path: modules/lib/picolibc
201       revision: 4f8076520240192be284843a114c92853e29bbbc
```
I cannot find this commit:  4f8076520240192be284843a114c92853e29bbbc  in https://github.com/zephyrproject-rtos/picolibc

enjiamai commented:
Update: it works after 0984aedf68e6ff967af93ee12ee7c5de81178ca6 got merged. thanks. Please ignore below comments, thanks!

keith-packard commented:
> Update: it works after [0984aed](https://github.com/zephyrproject-rtos/zephyr/commit/0984aedf68e6ff967af93ee12ee7c5de81178ca6) got merged. thanks. Please ignore below comments, thanks!

Sorry for messing up the zephyr/picolibc repo -- looks like we've got a process in place now with the ability to FF the main branch.
